### PR TITLE
Nexus memory: core persistence, retrieval, tools, and toggles

### LINKED ISSUE #1407: Nexus memory: core schema, retrieval injection, save/forget chat tools + capability
Part of #1406.

## Summary

The memory engine: schema, write-time safety pipeline, retrieval injection into the Nexus system prompt, `saveMemory`/`forgetMemory` chat tools, the `nexus-memory` capability, and the three off switches (global kill switch, per-user toggle plumbing, per-conversation toggle incl. UI). After this issue, memory works end-to-end via chat; the Settings surface comes in the next issue.

**MUST read `/docs/features/nexus-conversation-architecture.md` before touching conversation code (CLAUDE.md Critical Rule #6).**

### Schema — migration `162-nexus-user-memories.sql`

`nexus_user_memories`:
- `id uuid PK DEFAULT gen_random_uuid()`
- `user_id integer NOT NULL REFERENCES users(id) ON DELETE CASCADE`
- `content text NOT NULL` — **sanitized at write, always**
- `category text NOT NULL DEFAULT 'context'` — `profile` | `preference` | `context`
- `source text NOT NULL` — `tool` | `manual` | `auto` | `import:chatgpt` | `import:claude` | `import:gemini`
- `source_conversation_id uuid REFERENCES nexus_conversations(id) ON DELETE SET NULL`
- `embedding vector(512)`
- `deleted_at timestamptz` (soft delete)
- `created_at` / `updated_at` + `update_updated_at_column()` trigger

Indexes: HNSW `vector_cosine_ops` on `embedding` (partial `WHERE deleted_at IS NULL` — note migration 121 never added an HNSW index for text embeddings; this table adds its own) + btree on `user_id` partial on live rows. Raw SQL in `infra/database/schema/`, idempotent `IF NOT EXISTS`, **no `DO $$` blocks** (splitter limitation), registered in `infra/database/migrations.json` `migrationFiles`. Model on `157-rooms.sql`. Seed settings rows in the same migration (pattern: migration 120/121): `NEXUS_MEMORY_ENABLED` (true — global kill switch), `MEMORY_EMBEDDING_MODEL_ID` (`amazon.titan-embed-text-v2:0`), `MEMORY_EMBEDDING_DIMENSIONS` (512), `MEMORY_RETRIEVAL_THRESHOLD` (0.3), `MEMORY_RETRIEVAL_TOP_K` (6).

Drizzle model in `lib/db/schema/tables/nexus-user-memories.ts`, exported from `lib/db/schema/index.ts`.

### Memory service — `lib/nexus/memory/memory-service.ts`

- **Write pipeline (single choke point, used by every future write path):** `contentSafetyService.processInput` → embed via direct Bedrock helper (decoupled `MEMORY_EMBEDDING_MODEL_ID`, #1252 graph pattern — NOT the v1 `EMBEDDING_MODEL_*` pipeline) → dedup: cosine ≥0.90 against the user's live memories → UPDATE that row; otherwise INSERT. Embed **outside** any transaction (see comment at `lib/content/retrieval-service.ts:180`).
- **Retrieve:** embed the latest user message; return all live `profile`-category memories (always injected) + top-k (`MEMORY_RETRIEVAL_TOP_K`) `preference`/`context` memories with similarity ≥ `MEMORY_RETRIEVAL_THRESHOLD`. Fail-open: any retrieval error logs and returns empty — chat must never break because of memory.
- All operations scoped to the owning `user_id`; server actions/tools re-check `nexus-memory` capability + ownership on every call.

### Read-path injection

In `resolveToolsAndStream` (route.ts:2377): when global kill switch ON + capability granted + user toggle ON + conversation toggle not disabled → retrieve and pass a new `userMemoryFragment` parameter into `buildNexusSystemPrompt` (route.ts:1687), alongside the existing four fragments, with guidance text (what memories are, use naturally, don't recite).

### Chat tools

`saveMemory` / `forgetMemory` following `buildWorkspaceChatTools` (`lib/nexus/workspace-chat-tools.ts:612`): server-built from the resolved userId (never from client `enabledTools`), zod `inputSchema`, return `{tools, systemPromptFragment}`, merged in `buildMergedChatTools` (route.ts:199). Like attachment/repository tools, memory tools **bypass the skill `allowed-tools` pin** (documented choice, comment pattern at route.ts:229-232) — a bound skill must not silently disable a user's memory. Note: registering tools sets `multiStepToolsActive` → `maxSteps: 10` on turns that previously had no tools; existing Nexus E2E must stay green.

### Toggles

- Global: `NEXUS_MEMORY_ENABLED` setting checked in read + write paths (voice-mode kill-switch precedent).
- Per-user: honor `nexus_user_preferences.settings.memoryEnabled` (default true when absent) in read + write paths. UI for it ships with the Settings tab issue.
- Per-conversation: `nexus_conversations.metadata.memoryDisabled` (no migration needed); server honors it on both paths; Nexus UI control on the conversation (menu/composer) to flip it, persisted via existing conversation update path.

## Definition of Done

- [ ] Migration 162 creates `nexus_user_memories` (columns/indexes/trigger above) + settings seeds, is registered in `migrations.json`, and `bun run db:reset` applies it cleanly on local Docker
- [ ] Drizzle model exported; no `any`; no `console.*`
- [ ] `nexus-memory` capability in `lib/capabilities/manifest.ts` with defaultRoles administrator+staff+student; `bun run capabilities:generate` output committed; `capabilities:check` green
- [ ] Unit test proves every memory-service write passes content through `contentSafetyService.processInput` before storage (the epic's safety invariant)
- [ ] Unit test: dedup — a ≥0.90-similar save updates the existing row instead of inserting; a dissimilar save inserts
- [ ] Service/integration test: with a seeded memory, `buildNexusSystemPrompt` receives a `userMemoryFragment` containing it; with kill switch off / capability revoked / user toggle off / conversation toggle on, the fragment is absent and no memory tools are registered
- [ ] Unit test: memory tools reject operations on another user's memory rows
- [ ] E2E functional `nexus-memory-save-tool`: logged-in user tells the assistant to remember a fact → a `source='tool'` memory row for that user exists (poll via list action/API)
- [ ] E2E functional `nexus-memory-conversation-toggle`: toggle memory off on a conversation → send a message → no new memory rows and turn completes normally
- [ ] Existing Nexus E2E suite green (memory on and off) — no conversation-architecture regressions
- [ ] Full test suite green; zero new lint warnings in changed files; full lint exits successfully at baseline; typecheck clean

## Affected areas
- `infra/database/schema/162-nexus-user-memories.sql` (new), `infra/database/migrations.json`
- `lib/db/schema/tables/nexus-user-memories.ts` (new), `lib/db/schema/index.ts`
- `lib/nexus/memory/` (new: memory-service, memory-tools, fragment builder)
- `app/api/nexus/chat/route.ts` (fragment param + tool merge + gating)
- `lib/capabilities/manifest.ts` + generated catalog
- Nexus conversation UI (per-conversation toggle control)
- `tests/e2e/` (2 new functional specs)

## Out of scope
- Settings tab UI, per-user toggle UI (next issue)
- Import flow; auto-extraction (later issues)
- Any surface other than Nexus chat (Assistant Architect, Compare, agents)

## Risk & rollback
- Risk: medium — touches the Nexus chat hot path; mitigated by fail-open retrieval, non-fatal writes, and the documented invariants.
- Rollback: flip `NEXUS_MEMORY_ENABLED` off (no deploy) — chat behaves exactly as today; schema is additive.

### PR DESCRIPTION
**Pull request by krishagel**
## Summary

- add migration 162 and the Drizzle model for owner-scoped, soft-deletable Nexus memories with 512-dimensional Titan embeddings
- route all memory writes through content safety, reject PII-bearing durable writes, deduplicate at cosine similarity >= 0.90, and retrieve profile plus relevant preference/context memory
- inject untrusted recalled memory into the Nexus system prompt and add capability-gated `saveMemory` / `forgetMemory` tools; forgetting is restricted to owner-filtered IDs recalled in the current turn
- honor global, per-user, and per-conversation switches; add an optimistic conversation memory control
- register the `nexus-memory` capability and generated catalog entry

## Validation

- `bun run lint`
- `bun run typecheck`
- `bun run test:ci --runInBand` — 481 suites passed, 1 skipped; 4,440 tests passed, 15 skipped
- `bun run capabilities:check`
- `bun run build`
- fresh isolated PostgreSQL initialization applied migration 162 and verified columns, constraints, indexes, trigger, settings, cascades, and capability grants
- authenticated isolated Playwright conversation-toggle coverage passed (2 tests)
- all 4 new Playwright cases are discovered; live-provider cases remain gated by `E2E_RUN_EXTERNAL=1`
- exact `git range-diff` equality verified after rebasing onto the current `dev`

## Review follow-up

- resolved Claude Medium finding: detected PII and PII placeholders are rejected after the mandatory safety pass and before embedding/storage, avoiding both raw PII persistence and expiring token corruption
- resolved Claude Low finding: `forgetMemory` accepts only owner-filtered memory IDs surfaced in the current turn

## Notes

The live `saveMemory` and disabled-turn provider cases require explicit authorization to send synthetic E2E chat content to the configured external AI provider. No credentials or shared database were copied into this isolated worktree.

Closes #1407

### CODE REVIEW

**Comment by krishagel** [comment]
@claude Please perform a fresh, full review of this PR at exact head SHA 4b9d08a32610de25c3f4ea521daaedb043d5709b. Review the complete diff against issue #1407, its Definition of Done, AGENTS.md, and CLAUDE.md. Pay particular attention to the write-time safety choke point, user/conversation ownership, all four availability gates on reads and tool execution, pgvector dedup/retrieval SQL, prompt-injection boundaries, tool merging/maxSteps, migration correctness, and toggle rollback/capability visibility. Do not modify code. Post an explicit clean/no-findings verdict or concrete actionable findings on this PR.

**Comment by krishagel** [comment]
Resolved both findings in final candidate SHA `5e13e292f6146175e70220c48a7573550f94ccd7`:

- Medium: the required `processInput` pass still runs first, then writes with `hasPII` or a PII placeholder are rejected before ownership, embedding, or database work. Durable memory stores neither raw personal data nor expiring placeholders. Added a regression test proving no embedding/storage occurs.
- Low: `forgetMemory` now checks the requested UUID against the owner-filtered memories surfaced in the current turn before calling the owner-scoped soft delete. Added rejection and successful-bound-owner tests.

Full lint, typecheck, 4,440-test Jest suite, and production build pass. The two commits were rebased onto current `dev`; `git range-diff` reports both patches exactly equivalent.

**Comment by krishagel** [comment]
@claude Please perform the mandatory fresh, full final review at exact head SHA `5e13e292f6146175e70220c48a7573550f94ccd7` after the prior findings were fixed. Read the complete diff against current `dev`, AGENTS.md, CLAUDE.md, and the #1407 Definition of Done reproduced in the PR scope. Explicitly verify that (1) the mandatory content-safety pass is still first and detected PII/PII placeholders are rejected before ownership, embedding, or storage, so no raw PII or expiring token is persisted; (2) `forgetMemory` accepts only owner-filtered IDs surfaced in the current turn; and (3) the rest of the full safety, ownership, four-gate, pgvector, prompt-injection, tool-merging/maxSteps, migration, UI rollback, and test coverage remains correct after the rebase. Do not modify code. Post an explicit clean/no-findings verdict or concrete actionable findings.

**Comment by krishagel** [comment]
Additional isolated acceptance evidence on unchanged final SHA `5e13e292f6146175e70220c48a7573550f94ccd7`:

- Created a fresh issue-specific pgvector/PostgreSQL 16 tmpfs database on port 55407 and applied the full native fresh-install manifest through migration 162.
- Seeded the complete authenticated fixture set, including the unified-content repository, Atrium document/Yjs state, resource grants, and startup capability sync (`nexus-memory` granted to 3 default roles).
- Ran the complete non-provider Nexus scope against an issue-specific custom collaboration server: `bunx playwright test tests/e2e/nexus*.spec.ts tests/e2e/nexus/*.spec.ts --workers=1 --retries=0` with `E2E_EXCLUDE_EXTERNAL=1`.
- Result: **43 passed, 8 intentionally skipped live-provider/live-chat cases, 0 failed**.
- This specifically closes the two earlier fixture limitations: the authorized repository REST/MCP flow passed, and the real Atrium collaboration workspace panel edit passed. The deterministic memory-toggle tests also passed; the credentialed remember/disabled-turn provider cases remain explicitly gated.
- The custom server, tmpfs database, local Atrium storage, build cache, auth state, and Playwright artifacts were removed afterward; the worktree is clean and local/remote SHA still match.

**Comment by krishagel** [comment]
Final live-provider acceptance completed on unchanged final SHA `5e13e292f6146175e70220c48a7573550f94ccd7` using a fresh issue-specific PostgreSQL 16/pgvector tmpfs database (full manifest through migration 162), isolated custom server, and configured provider credentials:

`bunx playwright test tests/e2e/nexus-memory-save-tool.functional.spec.ts tests/e2e/nexus-memory-conversation-toggle.functional.spec.ts --grep "an explicit remember request is saved through the bound tool|a disabled conversation completes normally without writing memory" --workers=1 --retries=0`

Result: **2 passed, 0 failed**.

- Disabled conversation: the live provider turn completed normally and the owner-scoped marker query returned zero memory rows.
- Explicit remember request: the model invoked `saveMemory`; the test observed exactly one live owner-scoped row with `source = tool` and the exact source conversation id.
- Test cleanup returned the isolated database to zero live memory rows.
- Server, tmpfs database, build cache, auth state, local storage, and Playwright artifacts were removed. The worktree is clean and local/remote SHA still match.

With this final acceptance gate green, the PR has been marked ready for review.

**Comment by krishagel** [comment]
Rebased the same two issue #1407 patches onto current `dev` (`957d83bfff631aaedf7dc89dfe034127500e7887`). New final candidate: `6eb267c872e5a38a0bad7c60cb50c71c0a2e45dd`.

`git range-diff` reports both commits exactly equivalent to the previously reviewed/tested patches. Rebased local validation is green:
- `bun run capabilities:check`
- `bun run lint`
- `bun run typecheck`
- `bun run test:ci --runInBand` — 483 suites passed, 1 skipped; 4,473 tests passed, 15 skipped
- `bun run build`

The newly merged base commits do not overlap Nexus memory production or test paths; the only common generated file is the capability catalog, which remains synchronized. The prior authorized live-provider acceptance remains behaviorally covered by the exact-equivalent patch, while full build/test validation covers the new base.

@codex review

Please perform a fresh full review of exact head SHA `6eb267c872e5a38a0bad7c60cb50c71c0a2e45dd` against current `dev`, including the full issue #1407 scope and both prior review fixes. Post actionable findings or an explicit clean verdict.

File: lib/nexus/memory/memory-service.ts [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+          safety.blockedMessage || "This memory cannot be saved",
+          safety.blockedCategories || [],
+          "input",
+        )
+      }
+      const sanitized = safety.processedContent.trim()
+      if (
+        safety.hasPII === true ||
+        PII_PLACEHOLDER_PATTERN.test(sanitized)
+      ) {
```
[the remark this answers is not in the document]
Comment (krishagel) [reply lib/nexus/memory/memory-service.ts:128]: Fixed in 14d5e7220. PII tokenization now returns an explicit scanCompleted attestation, ContentSafetyService propagates it, and durable memory rejects every result other than piiScanCompleted === true before ownership lookup, embedding, or storage. Ordinary chat retains its existing fail-open behavior. Focused unit regressions cover disabled/failed/missing attestations and the clean path; the final live provider run completed Comprehend screening and persisted the tool memory.

File: lib/nexus/memory/memory-repository.ts [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+          .where(
+            and(
+              eq(nexusUserMemories.userId, userId),
+              eq(nexusUserMemories.category, "profile"),
+              isNull(nexusUserMemories.deletedAt),
+            ),
+          )
+          .orderBy(desc(nexusUserMemories.updatedAt)),
+      "listNexusProfileMemories",
+    )
```
[the remark this answers is not in the document]
Comment (krishagel) [reply lib/nexus/memory/memory-repository.ts]: Fixed in 14d5e7220. Profile retrieval now has a hard 20-row query limit, and the rendered system-prompt fragment has a 24,000-character ceiling that admits only complete JSON-quoted entries. Regression coverage proves the final fragment stays within budget and every included entry remains valid complete JSON.

**Comment by krishagel** [comment]
@codex review

Final review-fix SHA: `14d5e722016e3523f195b1abecc38e59a644eeff`.

Validation on this SHA:
- focused memory/safety suites: 52 passed
- full Jest: 483 suites passed, 1 skipped; 4,476 tests passed, 15 skipped
- `bun run lint`: passed
- `bun run typecheck`: passed
- `bun run build`: passed
- capability catalog check: passed
- authenticated memory toggle E2E: 2 passed, 2 expected live skips
- authorized live-provider acceptance run: 2 passed (disabled conversation writes no memory; explicit remember invokes `saveMemory` and persists exactly one owner-scoped tool row); disposable DB cleanup verified 0 live memory rows

Please review the final SHA, including the fail-closed PII attestation and bounded prompt-context fixes.

File: lib/nexus/memory/memory-service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (krishagel) [reply lib/nexus/memory/memory-service.ts:128]: Fixed in 14d5e7220. Durable memory writes now require an explicit piiScanCompleted === true result; a disabled or failing scanner returns false and saveMemory rejects before ownership lookup, embedding, or persistence. Ordinary chat remains fail-open. Focused safety/memory tests, the full Jest suite, and the authorized live Bedrock save-tool test all passed.

File: lib/nexus/memory/memory-repository.ts [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+          .select(selectedMemoryColumns())
+          .from(nexusUserMemories)
+          .where(
+            and(
+              eq(nexusUserMemories.userId, userId),
+              eq(nexusUserMemories.category, "profile"),
+              isNull(nexusUserMemories.deletedAt),
+            ),
+          )
+          .orderBy(desc(nexusUserMemories.updatedAt)),
```
[the remark this answers is not in the document]
Comment (krishagel) [reply lib/nexus/memory/memory-repository.ts]: Fixed in 14d5e7220. Profile retrieval is hard-capped at 20 rows per turn, and the final memory prompt fragment is capped at 24,000 characters while preserving only complete JSON entries. Regression coverage exercises both limits; focused tests, the full Jest suite, and production build passed.

File: lib/nexus/memory/memory-availability.ts [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  const [preference] = await executeQuery(
+    (db) =>
+      db
+        .select({ settings: nexusUserPreferences.settings })
+        .from(nexusUserPreferences)
+        .where(eq(nexusUserPreferences.userId, userId))
+        .limit(1),
+    "loadNexusMemoryUserToggle",
+  )
+  return preference?.settings?.memoryEnabled !== false
```
[the remark this answers is not in the document]
Comment (krishagel) [reply lib/nexus/memory/memory-availability.ts:69]: This control is intentionally owned by the next child issue. The #1407 acceptance criteria require honoring settings.memoryEnabled with a default of true and explicitly list the Settings-tab/per-user toggle UI as next-issue scope. #1408 explicitly owns the Memory tab and actions/nexus/memory.actions.ts setEnabled writer. This PR implements and tests the gate, including a seeded false value; adding the writer here would duplicate #1408 and expand this PR beyond its selected child. Resolving as out of scope for #1407.

File: app/api/nexus/conversations/[id]/route.ts [hunk 5]
```diff
@@ -60,44 +143,57 @@ export async function PATCH(
...
+          conversationId,
+          userId,
+        })
+        timer({ status: 'error', reason: 'not_found' })
+        return new Response('Conversation not found', { status: 404 })
+      }
+      updates.metadata = {
+        ...(existing.metadata ?? {}),
+        memoryDisabled,
+      }
```
[the remark this answers is not in the document]
Comment (krishagel) [reply app/api/nexus/conversations/[id]/route.ts]: Fixed in 69ef8b7a787a1328c01b88238c26f99e35b0817d. The PATCH route now uses one owner-scoped update with jsonb_set(COALESCE(metadata, {}), {memoryDisabled}, ...) and applies any combined title/archive/pin/save fields in that same statement, eliminating the stale metadata snapshot. A repository regression renders the Drizzle SQL and asserts the atomic merge and combined update. Validation passed: focused tests (17), full Jest (4,477 passed / 15 skipped), lint, typecheck, production build, authenticated toggle E2E (2 passed / 2 expected live skips), and authorized live Bedrock E2E (2 passed).

**Comment by krishagel** [comment]
@codex review

Please perform a fresh full review of exact head SHA 69ef8b7a787a1328c01b88238c26f99e35b0817d. The final change replaces the conversation memory toggle read-modify-write with an owner-scoped atomic jsonb_set update and adds SQL regression coverage. Local validation on this SHA: 4,477 Jest tests passed (15 skipped), lint, typecheck, production build, authenticated toggle E2E, and two authorized live Bedrock E2E cases.

**Comment by krishagel** [comment]
Follow-up to the full Claude review on 69ef8b7a7:

1. Concurrent dedup race: fixed in 7b9c4172fa814a8b0d0fedd45b8bb72bb098a1be. saveWithDedup now acquires a transaction-scoped advisory lock keyed by user before the nearest-row check, including the empty-set case. Regression coverage verifies the lock precedes the FOR UPDATE query; the real PostgreSQL/Bedrock save path also passed.
2. PII token scope: no durable memory can be written when PII is detected; any TTL token record is the existing content-safety behavior and is never read by memory. Accepted low-risk v1 behavior.
3. Sensitive non-PII content: accepted v1 boundary. Saves require an explicit model tool call under the user request, tool policy prohibits personal/sensitive material, and the backend requires a positive PII scan. Broader sensitivity classification is outside #1407.
4. Purge: #1407 explicitly specifies deleted_at soft delete for forgetMemory; physical purge is not part of this child.
5. Duplicate DI types and rollback-comment notes are non-blocking nits with no behavior impact.

Final local evidence: lint, typecheck, production build, 4,478 CI-equivalent Jest tests passed (15 skipped), authenticated toggle E2E passed, authorized live Bedrock E2E passed, and cleanup left zero memory rows.

@codex review
Please perform a fresh full review of exact head SHA 7b9c4172fa814a8b0d0fedd45b8bb72bb098a1be.

File: lib/nexus/memory/memory-repository.ts [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+    // external Bedrock call.
+    return executeTransaction(
+      async (tx) => {
+        // FOR UPDATE cannot lock a missing nearest row, so serialize the
+        // check-then-insert decision per user even when the memory set is empty.
+        await tx.execute(
+          sql`SELECT pg_advisory_xact_lock(
+            ${NEXUS_MEMORY_DEDUP_LOCK_NAMESPACE},
+            ${record.userId}
+          )`,
```
[the remark this answers is not in the document]
Comment (krishagel) [reply lib/nexus/memory/memory-repository.ts:134]: Fixed in 136221dbbffd0be38e129db0c63ebaa4d450ddef. The per-user advisory lock remains the serialization primitive, while the transaction now explicitly uses READ COMMITTED so the nearest-row statement takes a fresh snapshot after any lock wait and sees the prior committed insert. The regression asserts this isolation level. A direct real-PostgreSQL concurrency test launched two identical saveWithDedup calls simultaneously: results were one inserted and one updated with the same UUID, and the table contained exactly one row. Lint, typecheck, production build, all 4,478 CI-equivalent Jest tests, and both authorized live Bedrock cases passed.

**Comment by krishagel** [comment]
@codex review

Please perform a fresh full review of exact head SHA 136221dbbffd0be38e129db0c63ebaa4d450ddef. The final fix keeps the per-user advisory lock and changes the transaction to READ COMMITTED so the post-wait nearest-row query receives a fresh snapshot. Direct concurrent PostgreSQL proof: two simultaneous identical saves returned one inserted and one updated with the same UUID, with exactly one database row. Final local validation also includes lint, typecheck, production build, 4,478 CI-equivalent Jest tests, and both authorized live Bedrock cases.

File: lib/nexus/memory/memory-repository.ts [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        )
+        const literal = vectorLiteral(record.embedding)
+        const nearestResult = await tx.execute(sql`
+          SELECT id, 1 - (embedding <=> ${literal}::vector) AS similarity
+          FROM nexus_user_memories
+          WHERE user_id = ${record.userId}
+            AND deleted_at IS NULL
+            AND embedding IS NOT NULL
+          ORDER BY embedding <=> ${literal}::vector
+          LIMIT 1
```
[the remark this answers is not in the document]
Comment (krishagel) [reply lib/nexus/memory/memory-repository.ts:149]: Fixed in ef01a1a905f2b86d628931d0b97bc2d4b8ee35a9. Both dedup and relevant-memory retrieval now first materialize the complete owner-scoped live subset, then perform exact cosine ranking outside that CTE, preventing global HNSW candidate post-filtering from dropping owner rows. SQL regressions cover both paths. Direct PostgreSQL proof included an identical vector for another owner: two simultaneous same-owner saves collapsed to one UUID, and retrieval returned only the requested owner row. Lint, typecheck, production build, 4,479 CI-equivalent Jest tests, and both authorized live Bedrock cases passed.

**Comment by krishagel** [comment]
@codex review

Please perform a fresh full review of exact head SHA ef01a1a905f2b86d628931d0b97bc2d4b8ee35a9. Dedup and relevant retrieval now materialize the complete owner-scoped live subset before exact cosine ranking, preventing global HNSW post-filter loss. Direct PostgreSQL proof covered another owner with an identical vector, concurrent same-owner dedup to one UUID, and owner-only retrieval. Final local validation: lint, typecheck, production build, 4,479 CI-equivalent Jest tests, and both authorized live Bedrock cases.

File: lib/nexus/memory/memory-service.ts [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+      )
+    },
+
+    async retrieve(input) {
+      try {
+        const profilePromise =
+          dependencies.repository.listProfileMemories(
+            input.userId,
+            MAX_PROFILE_MEMORIES_PER_TURN,
+          )
```
[the remark this answers is not in the document]
Comment (krishagel) [reply lib/nexus/memory/memory-service.ts:173]: Fixed in d628ea8f278d5a49127b7ed479089acd69076b60. Bounded semantic retrieval now includes profile rows as well as preference/context, and the service deduplicates those results against the newest-20 profile set before prompt/tool binding. This keeps prompt bounds while allowing an older relevant profile ID to reach forgetMemory. Direct PostgreSQL proof seeded one old plus 20 newer profiles: the old row was absent from listProfileMemories(20) but returned by semantic retrieval. Lint, typecheck, production build, 4,479 CI-equivalent Jest tests, and both authorized live Bedrock cases passed.

**Comment by krishagel** [comment]
@codex review

Please perform a fresh full review of exact head SHA d628ea8f278d5a49127b7ed479089acd69076b60. Bounded semantic retrieval now includes profile rows and deduplicates them against the newest-20 profile set, preserving prompt limits while making older relevant profile IDs available to forgetMemory. Direct PostgreSQL proof showed an old profile absent from the newest 20 but returned semantically. Final local validation: lint, typecheck, production build, 4,479 CI-equivalent Jest tests, and both authorized live Bedrock cases.

File: lib/nexus/memory/memory-repository.ts [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+              content,
+              category,
+              source,
+              source_conversation_id,
+              embedding,
+              created_at,
+              updated_at
+            FROM nexus_user_memories
+            WHERE user_id = ${userId}
+              AND category IN ('profile', 'preference', 'context')
```
[the remark this answers is not in the document]
Comment (krishagel) [reply lib/nexus/memory/memory-repository.ts:253]: Fixed in e0efc6b873c76feeb3b1fc4025275911d09d92b2. The service now passes the exact newest-profile IDs into the owner-scoped semantic query, which excludes them with parameterized UUID predicates inside the materialized owner CTE before cosine filtering, ordering, and LIMIT. This preserves the full semantic top-K for older relevant profile/preference/context rows while retaining defensive service deduplication and profile-only fail-open behavior. Unit regressions assert the call contract and SQL ordering. A direct PostgreSQL proof reproduced six newest profiles consuming the control top-K, then returned the older profile plus preference/context after exclusion. Lint, typecheck, 4,479 CI-equivalent tests, production build, and both authorized live Bedrock cases passed.

**Comment by krishagel** [comment]
@codex review

Please perform a fresh full review of exact head `e0efc6b873c76feeb3b1fc4025275911d09d92b2`. This final fix excludes already-loaded newest profile IDs inside the owner-scoped materialized CTE before cosine ranking and semantic `LIMIT`, preserving top-K capacity for older relevant profile/preference/context rows. Validation on this SHA: lint, typecheck, focused 18 tests, full CI-equivalent Jest (484 passed suites; 4,479 passed tests; 15 skipped), production build, direct isolated PostgreSQL edge-case proof, and both authorized live Bedrock memory cases (2/2) passed; disposable DB cleanup confirmed zero rows.

File: app/api/nexus/chat/route.ts [hunk 10]
```diff
@@ -2459,6 +2462,8 @@ async function resolveToolsAndStream(params: {
       projectBinding: prepared.projectBinding,
       skillRepositoryIds: skillBinding.skillRepositoryIds,
     }),
+    memoryTools: memoryContext.tools,
```
[the remark this answers is not in the document]
Comment (krishagel) [reply app/api/nexus/chat/route.ts:2472]: Fixed in 7ceef3e0651798cdf24bc658b7fa450062c58e44. The validated model config now carries provider function-calling metadata, and memory binding reuses the shared router eligibility contract. Models that explicitly opt out—and Latimer regardless of stale/missing metadata—receive no pre-resolved saveMemory/forgetMemory tools, while owner-scoped recalled context remains in the system prompt as read-only facts. The shared router now also treats Latimer as function-call-ineligible for every required-tool route. Focused config/router/context regressions pass, including the read-only recall case. Lint, typecheck, all 4,481 CI-equivalent Jest tests, production build, and both authorized live Bedrock cases passed; disposable DB cleanup confirmed zero rows.

**Comment by krishagel** [comment]
@codex review

Please perform a fresh full review of exact head `7ceef3e0651798cdf24bc658b7fa450062c58e44`. The final compatibility fix suppresses pre-resolved memory tools for models that cannot execute function calls (including Latimer) while retaining read-only owner-scoped recalled context, and reuses the same eligibility helper in shared model routing. Validation on this SHA: 36 focused tests, lint, typecheck, full CI-equivalent Jest (484 passed suites; 4,481 passed tests; 15 skipped), production build, both authorized live Bedrock memory cases (2/2), and zero-row disposable DB cleanup.
